(ISSUE-21) Work on distribution of notty configuration data throughout library.

waynr opened the pull request:
@withoutboats here's what I've got so far.

Currently I am working to remove all usage of `Config::default()` as the source of configuration other than in `scaffolding` and the tests in various modules since I'm not sure how else we would make use of configuration with the approach I've taken so far of replacing `default()` methods with `new(config: Config)` for all structs that need to be passed configuration data.

The problem I am running into currently is what appears to be overgeneralization of the `Grid` class:
https://github.com/withoutboats/notty/blob/master/src/terminal/char_grid/grid.rs#L24

Looking through the rest of the source code I only see two specific types of `Grid`: [`Grid<CharCell>`](https://github.com/withoutboats/notty/blob/master/src/terminal/char_grid/mod.rs#L40) and [`Grid<i32>`](https://github.com/withoutboats/notty/blob/master/src/terminal/char_grid/grid.rs#L297). The reason I think of this as problematic is because there are a number of `T::default()` calls used to construct `Grid` elements--which sort of enforces the need for [`CharCell` structs to call `Config::default()` in `CharCell::default()`](https://github.com/waynr/notty/blob/issue-21/src/terminal/char_grid/cell.rs#L137) to obtain configuration data, which makes it difficult for me to imagine a way to pass the `Terminal`-owned Config down to the `CharCell`. Here are a couple possiblities I am considering:
- Create a `Configurable` trait that requires a `configure(&self, config: Config)` method, add that to the [`Grid` type constraints](https://github.com/withoutboats/notty/blob/master/src/terminal/char_grid/grid.rs#L34); then use this to configure the `CharCell` created by `T::default()`.
- Remove generics from `Grid` entirely since the only specific type used in this library is `Grid`. I think this is my preferred approach, unless you anticipate new specific types of `Grid` in the near future.

waynr committed:
Use Config member in Terminal.

* Initialized from Config::default()
* Then pass specific Config values down to TextRenderer from the Renderer
  instance of the terminal object.
* Implement 'Default' trait for Config, wrap CONFIG with that. This enables the
  systematic replacement of CONFIG with Config::default().

waynr committed:
Use Config member in CharGrid.

Initialized from Config::default().

waynr committed:
Use Config member in Cursor.

* Initialized from Config::default().
* Implement Clone on Config since Cursor derives Copy & Clone.
* Remove Default implementation on Cursor.
* Implement new method on Cursor.

waynr committed:
Use Config member in CharCell and Styles.

* Initialized from Config::default().
* Implement Debug and PartialEq for Config to facilite the addition of a Config
  member on the Styles struct, which derives these traits.
* Add "new" constructor for CharCell Styles in order facilitate
  initialization of their config members wherever they are used.
* Also remove the Default implementation for Styles since intialization requires
  access to a Config and the intent is to stop using the global static CONFIG
  object.

waynr committed:
Don't use Config::default in CharGrid initializer.

withoutboats commented:
Haha! This is a pretty interesting case, I'm glad it emerged because its a pretty fun problem to think about imo.

I think there are actually a few other options in addition to the two you mentioned, and I think any of them would be suitable _except_ the one you said you preferred - removing the type parameter from `Grid`. Let me explain why.

The most obvious advantage of generics is as you identified - that we can easily reuse a `Grid<T>` with a wide variety of different types. And you're 100% right that this advantage is unimportant for this use case - we use a `Grid<i32>` for testing, but in practice we only care about `Grid<CharCell>`.

Generics have another advantage, though, which I think is even more important: they enforce encapsulation. `Grid<CharCell>` knows everything there is to know about `CharCell`; `Grid<T>` only knows what we tell it about `T`. I'd much rather have the compiler enforce that I don't call arbitrary methods on `CharCell` than be responsible for enforcing that rule myself, and I'm willing to write a little bit of boilerplate to get this guarantee.

Currently, we have allowed `Grid` to expect that `T` is `Default + Clone`. But it turns out our `T` is not `Default` anymore, so what we need to do is rewrite `Grid` so it doesn't have that `Default` requirement anymore, while telling `Grid` as little as possible about `T` - `Grid` needs to know how to make a `T` when it resizes itself so that it can fill itself with 'default' `T`s, and there are a few ways we could communicate this to `Grid`:
- As you mentioned, we could replace `Default` with a new trait, which constructs a `T` from a `Config` struct, and implement that trait for both `CharCell` and `i32`. Note that I think the function defined in the trait should be a constructor, not a method (`fn(Config) -> Self` rather than `fn(&self, Config) -> Self`) This is a very valid option, but it raises the question: do we pass `Config` into every method that could create a new `T`, or do we store a `Config` on `Grid`? Either variant is a valid solution, and I don't feel strongly about the difference between them.
- We could also require a 'default' value to be passed to the `Grid` constructor, store that value on the `Grid`, and then clone it every time we need to create a new member of the grid. This would reduce the bounds on `T` to just `T: Clone`.
- Finally, every function that currently calls `T::default` could instead accept a function as an argument, with the bound `Fn() -> T`. For the `Grid<i32>` test cases, we would pass `|| 0` as this argument, and for the actual cases in `CharGrid` we would pass a closure like `|| CharCell::config(self.config)` or whatever. Something that is both an advantage and a disadvantage here is that every time you call a method, you can make a different decision about how it should construct a `T`.

Any of these are fine solutions in my opinion; there are differences between them in terms of performance, boilerplate, and flexibility, but I don't think they are significant (in particular any performance difference is essentially nil, don't worry about that at all). Whichever of these sounds best to you I'd be happy with.

withoutboats commented on src/cfg.rs:
```diff
@@ -25,6 +28,60 @@ pub struct Config {
+            fg_color: self.fg_color.clone(),
+            bg_color: self.bg_color.clone(),
+            cursor_color: self.cursor_color.clone(),
+            colors: self.colors,
```
Hah, I was wondering why you had a manual impl for `Clone`. I can't wait until Rust has const generics and we don't have to do silly things because arrays with lengths above 32 don't implement anything.

withoutboats commented on src/terminal/char_grid/cursor.rs:
```diff
@@ -24,10 +24,20 @@ pub struct Cursor {
     pub coords: Coords,
     pub style: Styles,
     pub text_style: Styles,
+    config: Config,
```
Instead of duplicating `Config` here, do you think you could add an argument to every function that needs that info on `Cursor`? There's a 1-to-1 relationship between `Cursor` and a `CharGrid` so there's no point duplicating this info.

waynr replied:
Sure

waynr replied:
@withoutboats what do you think about storing a reference to a `Config` instead?

withoutboats replied:
This will raise lifetime issues. Ultimately `Cursor` and `CharGrid` and so on are all owned `Terminal`, which also owns the `Config`. You can't store a reference into the same object.

~~I'll PR this branch with what I think is the best way to get around this problem.~~

withoutboats replied:
Scratch that last comment, the clever thing I wanted to do is actually unsound.

We can't store the config reference on the child structs, which means we have to pass it in as an argument to every function that needs it, like you've done with the renderers. This is a lot of boilerplate but its how it works. But wait - there's more!

So far, in order to be able to iterate quickly, I've had `Terminal` deref to `CharGrid`, so I didn't have to write delegation methods. Not sure if you know this, but when you call a method like `terminal.move_cursor`, Rust will insert as many dereference operators before `terminal` as it needs to do get to a type that has a `move_cursor` method.

However, now the signature of a lot of `CharGrid`'s methods will change to add the config argument. That means we need explicit delegations from `Terminal` to `CharGrid` which pass in `Terminal`'s config struct. Sorry, lot's of boilerplate.

waynr replied:
> This will raise lifetime issues. Ultimately Cursor and CharGrid and so on are all owned Terminal, which also owns the Config. You can't store a reference into the same object.

Okay, this makes sense, I didn't actually think it would work as phrased...I am still trying to understand the various options available here as far as Rust is concerned so I hope you don't mind if I bikeshed a little. Specifically, I was thinking that rather than copying the `Config` around everywhere or having it owned by `Terminal` and referenced elsewhere as `&Config` that we could do something like allocate new `Config`s on the heap using either `Rc<Config>` or `Arc<Config>`--if I understand correctly, the choice depends on whether we need the `Config` to be available in more than one context. (EDIT: i mean more than one threaded context)

My goal isn't just to avoid boiler plate but also to only pass references as a way to discourage copying and potentially storing configuration somewhere that would become a liability--if we have a UI menu somewhere that makes changes to a Config object, that change should propagate by virtue of that fact that we are changing a reference rather than explicitly calling an update method everywhere it is needed.

> Not sure if you know this, but when you call a method like terminal.move_cursor, Rust will insert as many dereference operators before terminal as it needs to do get to a type that has a move_cursor method.

I didn't realize this, although I knew such a thing was possible...I think you pointed it out at one point in one of my earlier PRs. So I guess if it's not possible/practical/desirable to give each struct a `Rc<Config>`, what do you think about storing the `Config` on the `CharGrid` instead and relying on the `Terminal` to deref to `CharGrid` to access it? Does that even make sense?

withoutboats replied:
> I was thinking that rather than copying the Config around everywhere or having it owned by Terminal and referenced elsewhere as &Config that we could do something like allocate new Configs on the heap using either Rc<Config> or Arc<Config>--if I understand correctly, the choice depends on whether we need the Config to be available in more than one context. (EDIT: i mean more than one threaded context)

I think it would make sense to put the Config in an Rc, and have the `Terminal` and each `CharGrid` have a separate instance of it, but to pass it by reference to each function that needs it on any of the components of `CharGrid`. Reference counting does have a cost, we don't want to have to pay it every time we create a new character cell, but creating new grids is a much less frequent thing.

I think `Rc` is fine instead of `Arc` here.

> what do you think about storing the Config on the CharGrid instead and relying on the Terminal to deref to CharGrid to access it? Does that even make sense? 

It's what I thought we could do, but it doesn't work.

Right now, the Terminal dereferences to the active CharGrid. So you can call `terminal.move_cursor(movement)` and the cursor in the active char grid will get moved. I thought we could have the Terminal dereference to some temporary struct that held references to the CharGrid and the Config, but that doesn't work because you can't dereference to something that didn't already exist.

waynr replied:
> I think it would make sense to put the Config in an Rc, and have the Terminal and each CharGrid have a separate instance of it, but to pass it by reference to each function that needs it on any of the components of CharGrid.

Cool, I'll go ahead and take this approach then.

withoutboats commented on src/terminal/char_grid/styles.rs:
```diff
@@ -29,9 +29,25 @@ pub struct Styles {
     pub strikethrough: bool,
     pub inverted: bool,
     pub blink: bool,
+    config: Config,
```
Similarly, I'd rather not duplicate the whole config struct for every `Styles` struct. Instead of keeping this around, pass the `Config` from the `CharGrid` as an argument to the `update` function. The `Config` struct is about 800 bytes large because of the color array, and there'll be one of these created for every cell of the grid. Currently the `Styles` struct is like 10 bytes or so. This difference will add up when you have possibly tens or even hundreds of thousands of cells in a terminal which keeps a scrollback history in multiple terminal tabs.

(Doing some napkin map - if you have a little over 300 chars of terminal space, which is not hard for a desktop monitor, then that's about 256k per line, meaning 4k lines of scrollback is a gigabyte of data, which would not work out!)

waynr replied:
My intent was originally to just get this working then later (in this same PR) once the `Config` is being initialized in `scaffolding` to experiment with memory optimization by keeping references to a per-`Terminal` config.

withoutboats replied:
Sounds good. There'll be challenges to storing `Config` on `Terminal` because of the way `Deref` is being used to dispatch methods to `CharGrid`.

waynr committed:
Store a default T value on Grid for cloning.

This removes the need for T to implement Default, which was problematic since
CharCell (a type of T) needs access to configuration data which should be passed
to it from the Grid.

waynr committed:
Initialize Styles using Config reference.

Rather than store a copy of Config on the Styles struct, make Config a required
parameter when initializing or updating Styles instances.

waynr committed:
Pass Config as a parameter to Terminal constructor.

waynr committed:
Make char_dimentions a method on Renderer.

waynr committed:
Make Config.font a String rather than a slice.

This is necessary to avoid always tying the lifetime of whatever the source of
the "font" setting to the lifetime of the Config object. Put another way, a
Config should always own its members.

* Remove Copy derivation for Config. Instead, explicitly clone Config where
  necessary. This is necessary because String objects are not and never will be
  Copy-able (for reasons I don't fully understand yet).
* Also remove cfg::CONFIG since it is never used anywhere else and we need to be
  able to call a function to set the 'font' member in Config::Default()

waynr committed:
Use Palette instead of [Color; 256].

Palette is a more palatable abstraction than [Color; 256]. Specifically because
it is a wrapper around a Vec<Color> which avoids the diffiulties involved with
[T;N] (no derivable traits for N > 32) while also ensuring that Palettes have
exactly 256 colors since Palette's 'colors' member is private--can only be
directly modified using Palette methods.

It's also worth mentioning that an abstraction like Palette ought to lead to
some interesting (but maybe ultimately useless) possibilities, like adding a
"color shift" parameter that would allow Colors returned from Palette::index()
to be modified by some factor whenever a Palette value is requested.

waynr committed:
Remove trailing whitespace.

waynr committed:
Initialized Config::default from toml string.

Store all Config defaults in a resources/default-config.toml and load that
during notty compile to define default Config values.

waynr committed:
Remove manual implementation of Debug for Config.

waynr committed:
Refactor Config loading code.

* Remove temporary empty_config function.
* Implement new_from_toml and new_from_file methods for Config.
* new_from_toml directly queries the toml::Value::Table to grab config data from
  the toml config file.
* new_from_file takes a file path, opens the file, then calls new_from_toml to
  build the Config

waynr committed:
Move Config helper functions to bottom of file.

waynr committed:
Implement ConfigError.

Initially, only wrap std::io::error::Error, but lay the groundwork for handling
toml::Parser error states more gracefully as well.

waynr committed:
Add Parse(String) to ConfigError.

And ensure that it is in use in the parse_toml method that wraps
toml::Parser::parse() in a way that should really be handled by the toml library
itself.

waynr committed:
Add test for Config::new_from_file.

waynr committed:
Add test for Config::new_from_toml.

waynr committed:
Add test for Config::default().

waynr committed:
Implement Config::update_from_file() method.

This method allows Config struct values to be overridden by values set in a
toml configuration file specified by the "path" parameter. This commit also adds
a test to validate this behavior with a simple config file fixture in the
resources directory. I assume this file will probably need to be moved elsewhere
but will await PR feedback before doing so.

waynr committed:
Respect the SHELL environment variable, default to 'sh'

waynr committed:
Unlock runtime configuration achievement.

Use Config::default() but update from ~/.scaffolding.rc if it exists.

waynr committed:
Handle case where ~/.scaffolding.rc does not exist.

waynr commented:
@withoutboats so this is pretty much ready for review, although it needs a little more cleanup to be completely ready for merge. Left to do:
- [x] Always pass and store `Config` objects by reference.
- [ ] Write some more `Config` test cases to validate behavior under various error conditions.
- [ ] Move `Palette` definition to somwhere within `notty`...? It's not clear to me from the current state of discussion on withoutboats/notty-encoding#3 whether you still think it should be moved here. I personally still think it' would be useful for primarily ANSI-based apps to tap into a small part of the `notty` protocol to define an application-specific `Palette` but don't care enough at this point to keep arguing.
- [ ] Move the bg, fg, and cursor color definitions into the `Palette` struct.

withoutboats commented:
I'll review this either tomorrow or Thursday, thanks!

withoutboats commented on scaffolding/src/main.rs:
```diff
@@ -57,7 +58,12 @@ fn main() {
     env::set_var("TERM", "notty");
-    let (tty_r, tty_w) = tty::pty("sh", COLS as u16, ROWS as u16);
+
+    let shell = match env::var("SHELL") {
```
:+1:

withoutboats commented on src/cfg.rs:
```diff
@@ -1,314 +1,256 @@
-};
+impl Config {
+    /// Constructs a new Config from a toml file specifed by the input string "path".
+    pub fn new_from_file(path: &String) -> Result<Config> {
```
The signature of this function should be `new_from_file<P: AsRef<Path>>(path: P) -> Result<Config>`, and so should `read_toml_file`  and `update_from_toml` (see `std::fs::File::open` as an example). This signature means that you can pass a `String`, a `&str`, a `Path`, a `PathBuf`, whatever. It will also let you avoid a lot of type conversion methods elsewhere where you call this.

waynr replied:
Ah, interesting. Will definitely make this change. I have some questions about this but my brain is ready to crash for the night so I can't tell if they make sense.

withoutboats commented on scaffolding/src/main.rs:
```diff
@@ -70,7 +76,25 @@ fn main() {
+    // For now we don't care why the config failed to update, but in the
+    // reasonably near future we should figure out a more graceful way to
+    // detect and report configuration errors.
+    match config.update_from_file(&user_config_path
```
Per the note in the config file, you can just remove all the methods on this once you have a generic signature on these config methods.

withoutboats commented:
This looks pretty good. I made an inline comment about a way you can avoid some type conversions. There's one larger thing I'd like to change, sorry: I think the TOML parsing should happen in `scaffolding`, and not in `notty`. Implementations other than `scaffolding` may want to store configs in a totally different way, and I don't want them to have to have the dependency on toml that they don't use. I also think of `notty` as the "reference implementation" for a kind of terminal that will need to be copied widely if its to be actually useful, and storing configs in toml definitely isn't a part of the 'notty standard.'

This is kind of annoying I know because you can't implement this stuff as methods directly on the config struct if they're in `scaffolding`. :frowning:

> It's not clear to me from the current state of discussion on withoutboats/notty-encoding#3 whether you still think it should be moved here. I personally still think it' would be useful for primarily ANSI-based apps to tap into a small part of the notty protocol to define an application-specific Palette but don't care enough at this point to keep arguing.

Yeah, I'm sorry, I let the conversation drift into a more philosophical direction. If you think there are useful commands that can be implemented  using a `Palette` argument, I'll look forward to those proposals. For now I'll review that PR and we'll store `Palette` in the encoding library in preparation for that. Is it ready to land?

waynr commented:
> I think the TOML parsing should happen in scaffolding, and not in notty. Implementations other than scaffolding may want to store configs in a totally different way, and I don't want them to have to have the dependency on toml that they don't use.

Is it practical to hide the `toml`-using code behind [a conditional compilation flag ](https://doc.rust-lang.org/reference.html#conditional-compilation)? I don't have a real strong objection to moving most of the `Config` code, but I also don't see any reason not to provide hypothetical users of `notty` with the convenience of a means to read configuration values if there is a way they can easily opt out and provide their own config file reading code. Actually, my strongest objection is simply that I don't have much time..

I have considered factoring out most of the `Config` code here into a separate library that would provide macros that allow most of it to be generated as a mini "configuration DSL". This approach probably wouldn't be possible if we split the definition of the `Config` struct away from the definition of the methods that load it.

> I also think of notty as the "reference implementation" for a kind of terminal that will need to be copied widely if its to be actually useful, and storing configs in toml definitely isn't a part of the 'notty standard.'

It's not clear to me how the use of `toml` in the reference implementation makes it part of the "notty standard". If that's the case, would you also expect any application that uses `notty` as a reference to also us a data structure exactly like the `Config` to manage their own configuration? And to construct the other object relationships exactly as is done here? I would think that the important part of this library as a reference implementation of the `notty` protocol is the way it renders and passes input/output between the user and the application(s). The use of toml vs ini vs yaml vs json seems like a relatively trivial matter.

> This is kind of annoying I know because you can't implement this stuff as methods directly on the config struct if they're in scaffolding.

My understanding is that you can implement methods on items even if you didn't originally define those items. So if we do decide to move the toml code into `scaffolding` (which it looks like we will at this point) then it shouldn't be a problem to move the `impl` blocks also

> For now I'll review that PR and we'll store Palette in the encoding library in preparation for that. Is it ready to land?

Let me take a look tomorrow after work and get back to you, I'm feeling pretty droopy right now.

waynr committed:
Use static dispatch on the AsRef<Path> trait for Config methods.

waynr commented:
@withoutboats so I gave it a little more thought and, considering the relatively immaturity of the `toml` library I agree that it would be better to move that to a higher-level library. I still don't necessarily understand why implementation details of `notty` have the same kind of protocol-defining importance of `notty-encoding` (or why the latter is signficant/separate if this is the case). I'm not sure exactly when I will move the code, maybe this weekend.

waynr committed:
Move Config initialization & error handling to scaffolding.

waynr committed:
Remove extraneous functions.

These were only necessary before when loading the default configuration in
Config::Default()

waynr committed:
Remove extraneous tests.

waynr commented:
> I think it would make sense to put the Config in an Rc, and have the Terminal and each CharGrid have a separate instance of it, but to pass it by reference to each function that needs it on any of the components of CharGrid.

Okay, I think I am coming around to your way of thinking. In fact, I would say that we shouldn't even pass the entire `Config` reference to the `Cursor` method, just call the method on `Cursor` that sets its color whenever appropriate. I'll try to finish this part tonight or tomorrow night.

> So if we do decide to move the toml code into scaffolding (which it looks like we will at this point) then it shouldn't be a problem to move the impl blocks also

So I was wrong about this of course, it is not possible to `impl` methods on a struct defined in other libraries unless `impl`ing a particular trait on that struct. I think someone tried to explain this to me a couple weeks ago, too bad I don't learn just by having people say words at me...

As you can see, I have moved all the `toml` -> `Config` code into `scaffolding` in the past few commits. After doing this, I have somewhat of a concern about the `Config` object. Specifically, this `toml` -> `Config` code is directly dependent on every member of `Config` since it is responsible for populating new ones from compatible serialized types. I see this as a problem because now `scaffolding` is dependent on `Config` in a way that it was not before. If someone were to change the name or type of a `Config` member, it would be a backwards-incompatible change requiring a version bump in `notty` to avoid `scaffolding` accidentally upgrading to an incompatible minor or patch level version.

This means, for instance, that the comment in withoutboats/notty-encoding#3:

> Also, if you're going to encapsulate the 'config colors' into a single type, I think you should include the default foreground and background colors as well.

ought to imply a version bump in `notty` itself.

waynr commented:
So TL;DR for my last comments...I think I would like to hard reset this branch to 0a78bbf because it unnecessarily binds `scaffolding`, a user of `notty`, to an implementation detail of `notty`-specific Config.

waynr commented:
> Okay, I think I am coming around to your way of thinking. In fact, I would say that we shouldn't even pass the entire Config reference to the Cursor method, just call the method on Cursor that sets its color whenever appropriate.

waynr commented:
> Okay, I think I am coming around to your way of thinking. In fact, I would say that we shouldn't even pass the entire Config reference to the Cursor method, just call the method on Cursor that sets its color whenever appropriate.

This statement doesn't make sense, what I meant is "we shouldn't even pass the `Config` reference to the `Cursor` methods, instead just pass the necessary member values".

withoutboats commented:
> As you can see, I have moved all the toml -> Config code into scaffolding in the past few commits. After doing this, I have somewhat of a concern about the Config object. Specifically, this toml -> Config code is directly dependent on every member of Config since it is responsible for populating new ones from compatible serialized types. I see this as a problem because now scaffolding is dependent on Config in a way that it was not before. If someone were to change the name or type of a Config member, it would be a backwards-incompatible change requiring a version bump in notty to avoid scaffolding accidentally upgrading to an incompatible minor or patch level version.

Once notty is versioned, any breaking change to its public API requires a version bump, whether or not scaffolding depends on that part of the public API. Other, unknown dependencies (including those created in the future) need to be able to properly declare which versions of notty they're compatible with. notty is in exteme flux right now and so nearly every significant commit changes its public API.

While things are so unstable, scaffolding is only distributed with the source of notty, and the build_test script checks that scaffolding is in sync with the current public API (all commits merged into master _should_ pass build_test, though I've been lazy and that hasn't always been the case).

Notty also exposes a lot right now, like very few of the methods on `Terminal` or `CharGrid` are actually things I think should be public, its a byproduct of how I structured it so I could iterate quickly while things were still changing so much.

> This statement doesn't make sense, what I meant is "we shouldn't even pass the Config reference to the Cursor methods, instead just pass the necessary member values".

This makes a lot of sense, _especially_ when its only 1 member of config that's passed in.

waynr commented:
> Once notty is versioned, any breaking change to its public API requires a version bump, whether or not scaffolding depends on that part of the public API. 

This is true enough, I guess the reason I focused so strongly on this new tight binding between `scaffolding` and `Config` members is that `scaffolding` doesn't really have any reason to be exposed to details of `Config` except for the seemingly arbitrary constraint that `notty` not use `toml` directly. Maybe this will change someday in the future with something like a GUI color chooser? But even then, I wonder if it would be better design for the GUI to communicate using the `notty` protocol to set colors rather than directly editing `notty`'s configuration. Can you think of any reasons that `scaffolding` or any other hypothetical user of `notty` would need access to details of `notty::cfg::Config`? 

Anyway, I am beginning to annoy myself with my argumentativeness, I hope I'm not annoying you also :). You didn't directly respond to the preference I expressed for resetting this branch back to 0a78bbf--are you open to this?

> Notty also exposes a lot right now, like very few of the methods on Terminal or CharGrid are actually things I think should be public, its a byproduct of how I structured it so I could iterate quickly while things were still changing so much.

So this makes me wonder: is `notty` stable enough for a community of contributors and users? For example, when could I reasonably expect to write my own `notty` application or terminal and file bugs here when crates are released that make changes to the public API?

waynr committed:
Use Rc<Config> on Terminal, CharGrid, and Cursor.

waynr committed:
Remove Rc<Config> from Cursor.

waynr commented:
Just updated `Terminal` and `CharGrid` to use `Rc<Config>`.

waynr committed:
Remove unnecessary default-config.toml

waynr committed:
Merge remote-tracking branch 'origin/master' into issue-21

withoutboats commented:
Hey, sorry for not responding to some of your more substantial comments for a few days. Been allocating my notty time to the sub-dividable screens branch #9, which is approaching review time. :smile:

I've been having a lot of back and forth about how exactly configuration should be layered. Its complicated because it cuts through so many of the layers that exist right now (as you've certainly noticed, having to put config references into all these methods). I think right now the important thing is getting any sort of dynamic configuration landed, and we can iterate on it over time if we have to.

The upshot of this is that I'm fine with stepping back to the previous commit or anything else that you think makes sense. For that question, and for where Palette lives, I'll leave up to you. So I think that means the main thing left on this branch is testing?

It'd be exciting to get both this and #9 landed in the next week or so! I'll ping you when #9 could use a review.

waynr commented:
Rebasing this onto master is not trivial--I will hopefully feel like sitting at my computer sometime this weekend to work on it. Luckily all the commits in this branch were building and passing tests before I moved on to the next when I last worked on it.

withoutboats commented:
Hey, I'll be happy to handle the rebase if you want. Thanks for working on this.

withoutboats commented:
This has been squashed and merged, thanks for the PR!
